Source/WebKit/UIProcess/API/glib/WebKitImage.cpp

```diff
@@ -0,0 +1,477 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+ * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "WebKitImage.h"
+
+#include "WebKitImagePrivate.h"
+
+#if USE(SKIA)
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_BEGIN
+#include <skia/core/SkData.h>
+#include <skia/core/SkImage.h>
+#include <skia/core/SkImageInfo.h>
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN // Skia port
+#include <skia/encode/SkPngEncoder.h>
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
+WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_END
+#endif
+#include <span>
+#include <wtf/Assertions.h>
+#include <wtf/Hasher.h>
+#include <wtf/glib/GRefPtr.h>
+#include <wtf/glib/GUniquePtr.h>
+#include <wtf/glib/WTFGType.h>
+
+static void webkit_image_gicon_interface_init(GIconIface*);
+static void webkit_image_gloadable_icon_interface_init(GLoadableIconIface*);
+
+struct _WebKitImagePrivate {
+    int width;
+    int height;
+    unsigned stride;
+    CompletionHandler<GRefPtr<GBytes>(WebKitImage*)> createBytesFn;
+
+    GRefPtr<GBytes> bytes;
+};
+
+/**
+ * WebKitImage:
+ *
+ * This class represents a WebKit image.
+ *
+ * At the moment, this assumes native images are in RGBA format, with 4 bytes per pixel.
+ * When instantiating this class, stride must be large enough to accommodate the image width times
+ * the bytes-per-pixel representation.
+ * When instantiating this class and providing a CompletionHandler to instantiate the bytes, make sure
+ * this uses #CompletionHandlerCallThread::AnyThread if you first instantiate the class bytes asynchronously.
+ *
+ * Since: 2.52
+ */
+WEBKIT_DEFINE_FINAL_TYPE_WITH_CODE(
+    WebKitImage, webkit_image, G_TYPE_OBJECT, GObject,
+    G_IMPLEMENT_INTERFACE(G_TYPE_ICON, webkit_image_gicon_interface_init)
+    G_IMPLEMENT_INTERFACE(G_TYPE_LOADABLE_ICON, webkit_image_gloadable_icon_interface_init))
+
+
+enum {
+    PROP_0,
+
+    PROP_WIDTH,
+    PROP_HEIGHT,
+    PROP_STRIDE,
+
+    N_PROPERTIES
+};
+
+enum class ImageFormatBPP : uint8_t {
```

Comment:
BPP is typically an abbreviation for “**bits** per pixel”, but here is used as “**bytes** per pixel”. This is a bit confusing.

Also, this could be a `static constexpr`:


```suggestion
static constexpr auto RGBA8BytesPerPixel = 4;
```

Reply:
I thought about the `static constexpr` but chose the enum in case we added more -- since `GraphicsContextGL` defines so many formats. But I agree -- the constexpr covers what we need